### ISSUE TITLE
Run code line by line

### ISSUE THREAD

**Issue by liu-genomics** [vatlab/jupyterlab-sos#36]
I am trying to develop some scripts in Jupyterlab-sos so I need to run and test code line by line.  I have two blocks of code here. The first one is global variable definition, second one is something I need to run in bash.

![image](https://user-images.githubusercontent.com/16808441/60476874-ca5fda00-9caf-11e9-93d4-85c9454c6a35.png)
![image](https://user-images.githubusercontent.com/16808441/60476896-db105000-9caf-11e9-9b5d-012a7110187a.png)

When I ran line by line. It seems the variables were not recognized. How should I deal with this situation? I used ctrl+shift+enter to run line by line in a jupyterlab-sos browser interface.

![image](https://user-images.githubusercontent.com/16808441/60476933-02ffb380-9cb0-11e9-819e-82830b78d678.png)

![image](https://user-images.githubusercontent.com/16808441/60476940-0bf08500-9cb0-11e9-90a9-5347329e5e00.png)

**Comment by BoPeng** [comment]
Did you run the `east_snv` line in `[global]` section? I would usually step through the parameter section to get the parameter initialized

![image](https://user-images.githubusercontent.com/9889312/60478903-9f429f80-9c49-11e9-888e-0a5525afb55b.png)

**Comment by gaow** [comment]
@BoPeng I confirm it works using a MWE simlar to @liu-genomics posted where codes live in different cells:

<img width="1047" alt="Screen Shot 2019-07-02 at 10 01 08 AM" src="https://user-images.githubusercontent.com/917985/60523484-6b4e9500-9cb0-11e9-8d59-76fe19dc2f89.png">

However as you can see the `print(_input)` statement did not quite work.

**Comment by gaow** [comment]
Also, see below, line-by-line execution does not seem to work for scripts inside `bash` action

<img width="932" alt="Screen Shot 2019-07-02 at 10 03 26 AM" src="https://user-images.githubusercontent.com/917985/60523676-c1233d00-9cb0-11e9-945b-ce7920ab6b5a.png">

the first `ls` does not list the proper input (should be a single file not a number of them). The 2nd `echo` did not work. I've not used this feature myself to interactively develop codes inside actions in SoS Notebook so not sure if it is user error or something we do not support?

**Comment by BoPeng** [comment]
The first problem is https://github.com/jupyterlab/jupyterlab/pull/6606 , which might or might not be accepted because its negative impact on indent-less languages such as R. Right now you will have to manually select the entire `bash` block, or use classic jupyter for which I have implemented 6606.

The `_input` problem exists because `_input` is only defined when the workflow is executed with a `input:` statement so you will have to execute the entire cell or at least include the `input:` line with your selection. 

I could make `_input` persistent (e.g. after you step through an input statement), but think the case if you are running content with `_input` in another cell.

**Comment by gaow** [comment]
> I could make _input persistent (e.g. after you step through an input statement), but think the case if you are running content with _input in another cell.

Not sure if I understand this statement. In the context of interactive development we always would expect running the current cell based on content previously executed -- that case I've not seen an issue with persistent `_input`. Would you elaborate with an example?

> The first problem is jupyterlab/jupyterlab#6606 , which might or might not be accepted because its negative impact on indent-less languages such as R

Okay let's wait for the pull request on JupyterLab's end. In the mean time @liu-genomics if you want to do that within scripts you've got to fall back to `jupyter-notebook` based GUI.

**Comment by BoPeng** [comment]
>  Would you elaborate with an example?

The idea is that `_input` is supposed to be a transient variable only seen inside a step, and can change for each substep. So suppose you step through

```
input: 'a.txt', 'b.txt', group_by=1
```
SoS could ignore the line because there is no body of substep, or it could be the last `_input`, which should be  `file_target('b.txt')`.

Currently `_input` is reset right after you step through the line so that

```
print(_input)
```
would fail or print `` for an empty `sos_targets` (not sure which is the case).

You are right that we usually will not jump between the cells and use `_input` of one cell in another, but I am not quite comfortable to pass a hidden variable around.

**Comment by BoPeng** [comment]
You could comment on  jupyterlab/jupyterlab#6606 if you have some idea what the behavior should be.

**Comment by gaow** [comment]
> You are right that we usually will not jump between the cells and use _input of one cell in another, but I am not quite comfortable to pass a hidden variable around.

I agree with you that in SoS it does not make sense to access `_input` in a cell separate from where it is defined. However it seems this constraint has the side effect to line-by-line execution as I've shown above. Removing this contraint would then make this variable accessible by any other cells in the notebook, something not harmful given the context (my opinion) but is uncomfortable (your view). Is
 there a middle ground to this? 

BTW @liu-genomics when I develop workflow using SoS Notebook I often use a language specific kernel to develop the chunk of code in that language; then after it is confirmed to work I cut-paste those into sos cell putting in the workflow bells and whistles. It is slightly more tedious than what you are trying to do (developing all in SoS cell) but to me it is cleaner in a sense.

**Comment by liu-genomics** [comment]
Thanks very much Bo and Gao.  @gaow Could you please explain a bit more what do you mean by "when I develop workflow using SoS Notebook I often use a language specific kernel to develop the chunk of code in that language".  Do you also have variables in this context? For example, in shell script, we could write a `test.sh` with only one line of code `ls $1`. If I need to develop such code line-by-line in SoS Notebook, how should I write?

**Comment by gaow** [comment]
> Do you also have variables in this context? 

No. I would hard code input data variables etc to a separate language kernel, test the code make sure it works, then reformat that cell to an SoS workflow step -- this is how I usually work with Python and R; and in my view it is quite straightforward. For developing bash scripts I do use variables but I did not realize it does not work with JupyterLab -- last time I did it was long ago using Jupyter Notebook. I'm afraid for that purpose you've got to use Jupyter Notebook until they accepted our PR.

**Comment by BoPeng** [comment]
![image](https://user-images.githubusercontent.com/9889312/60533567-87a8fc80-9cc5-11e9-88de-65c9fdcd4da9.png)

Right now in sos you can use parameter `args` to specify parameters to shell (and other) actions, and SoS Notebook in classic jupyter can pick up the entire shell script. The behavior in JupyterLab is different in that line by line execution will only pick up the first line after the `sh` line, but hopefully jupyterlab/jupyterlab#6606 will be accepted soon.

**Comment by liu-genomics** [comment]
Thanks! Hopefully Jupyterlab would accept 6606 soon. Would love to have line-by-line execution with variables.

**Comment by BoPeng** [comment]
@gaow It is possible to retain `_input` after processing `input:` statement, but how about `output:`? 
```
output: _input.with_suffix('.bak')
```
might work all right, but
```
output: whatever, group_by=1
```
would fail miserably because there is no information on number of substeps. I guess the bottomline is that 
```
output: whatever, group_by=1
```
would not work in the current implement either so it is not going worse.

**Comment by gaow** [comment]
Well, at least in my workflows outputs are mostly derived from inputs. I think it is good enough to support this scenario. We did not stress heavily on line-by-line development for SoS Workflows but it is good to see there is a need for it.

**Comment by BoPeng** [comment]
I have pushed some changes to allow persistent `step_input`, `_input` etc. It roughly works but as you can see, there will be a warning message saying `_output` does not exist if you step through `output` statement.

![image](https://user-images.githubusercontent.com/9889312/60602820-2858f400-9d7a-11e9-9f3f-d80c11e3d1cb.png)

**Comment by BoPeng** [comment]
Unfortunately jupyterlab/jupyterlab#6606 was rejected so the behavior will be different between classic Jupyter and JupyterLab unless we change the behavior on the classic jupyter end..

**Comment by BoPeng** [comment]
I think this also gives us an opportunity to fine tune the behavior. That is to say, we can add a property in our language module and limit the "full block" behavior only to Python and SoS.

**Comment by gaow** [comment]
@BoPeng is the line-by-line execution feature broken? I cannot seem to get it work now with latest JupyterLab, although Jupyter Notebook works:

![Screenshot from 2019-10-02 08-38-39](https://user-images.githubusercontent.com/917985/66048929-0f361f80-e4f0-11e9-8027-78ee2fd46aa1.png)

when I hit ctrl shift enter nothing happens.

**Comment by BoPeng** [comment]
That is now a jupyterlab feature. I think I enabled the shortcut but perhaps I did not. Check if run in console from menu works.

**Comment by gaow** [comment]
Sorry I cannot seem to find that "run in console" in menu ... yes it would be nice if we are consistent with Jupyter Notebook IDE for SoS. I'm on the current pip release of Jupyterlab version 1.1.4

**Comment by BoPeng** [comment]
It appears to be working, but not at first when the SoS kernel has not been started, which by itself could be a bug.

**Comment by benbenz** [comment]
I am not sure this is related but I've noticed a very similar issue. It is also visible at [http://sosworkflows.com/](http://sosworkflows.com/)

If you start a new SoS console and type `%use Julia` for example, nothing will happen. The error is as follow:
```
18:05:08.664 › TypeError: Cannot read property 'model' of null
    at new v (http://localhost:8888/lab/extensions/jupyterlab-sos/static/5.e9a19b6a8ff02c8ed090.js?v=e9a19b6a8ff02c8ed090:1:9450)
    at S.get_info (http://localhost:8888/lab/extensions/jupyterlab-sos/static/5.e9a19b6a8ff02c8ed090.js?v=e9a19b6a8ff02c8ed090:1:11919)
    at http://localhost:8888/lab/extensions/jupyterlab-sos/static/5.e9a19b6a8ff02c8ed090.js?v=e9a19b6a8ff02c8ed090:1:31935
...
```

SOLUTION: You have to create an empty SoS notebook then create the SoS console and it should work.

It seems to be related to the construction of default NotebookInfo while the notebook passed is being null (?)
line 200 of manager.ts in jupyterlab-sos extension code:

```
public get_info(notebook: NotebookPanel): NotebookInfo {
    if (!this._info.has(notebook)) {
      console.log("Creating a new notebook info");  
      this._info.set(notebook, new NotebookInfo(notebook));  
    }
    return this._info.get(notebook);  
  }
```

 into line 44 of manager.ts:

```
if (notebook.model.metadata.has("sos"))
```

**Comment by benbenz** [comment]
> @BoPeng is the line-by-line execution feature broken? I cannot seem to get it work now with latest JupyterLab, although Jupyter Notebook works:
> 
> when I hit ctrl shift enter nothing happens.

I am not sure this is related the main issue here but I've noticed a very similar issue to the one @gaow is seeing. It is visible at [http://sosworkflows.com/](http://sosworkflows.com/)

If you start a new SoS console and type `%use Julia` for example, nothing will happen. When running locally, I also see the following error:
```
18:05:08.664 › TypeError: Cannot read property 'model' of null
    at new v (http://localhost:8888/lab/extensions/jupyterlab-sos/static/5.e9a19b6a8ff02c8ed090.js?v=e9a19b6a8ff02c8ed090:1:9450)
    at S.get_info (http://localhost:8888/lab/extensions/jupyterlab-sos/static/5.e9a19b6a8ff02c8ed090.js?v=e9a19b6a8ff02c8ed090:1:11919)
    at http://localhost:8888/lab/extensions/jupyterlab-sos/static/5.e9a19b6a8ff02c8ed090.js?v=e9a19b6a8ff02c8ed090:1:31935
...
```

I found this solution: create an empty SoS notebook then create the SoS console, and it should work.

This seems to be caused by `Manager.currentNotebook` being null in jupyterlab-sos labextension.

- line 1007 of `index.ts`:

```
      labconsole.promptCellCreated.connect(panel => {
        let info = Manager.manager.get_info(Manager.currentNotebook);
        addLanSelector(panel.promptCell, info);
      });
```
- into line 200 of `manager.ts`:

```
public get_info(notebook: NotebookPanel): NotebookInfo {
    if (!this._info.has(notebook)) {
      console.log("Creating a new notebook info");  
      this._info.set(notebook, new NotebookInfo(notebook));  
    }
    return this._info.get(notebook);  
  }
```

- into line 44 of `manager.ts` (with `notebook=null`):

```
if (notebook.model.metadata.has("sos"))
```

**Comment by BoPeng** [comment]
Thanks for the report. I will have a look.

**Comment by benbenz** [comment]
Great ! Thanks. And thank you for making such a great extension like SoS.
I just realized the console was probably meant to be associated with a notebook. But, as one can create a standalone console from the launcher, it would be great if it worked autonomously. I tried to modify the code to create a hidden NotebookPanel or separate the formalism betweeen notebooks and consoles (more involved in terms of code) with no luck so far. It's hopefully an easy fix for you :)

**Comment by BoPeng** [comment]
I pushed a simple fixes and the console appears to be working. However, because the console cannot communicate with the backend to get a list of kernels, it does not have the language dropdown so the console is not fully operational.

I suppose with some more work we can get the language dropdown working for this case but maybe it does not worth the trouble.

**Comment by benbenz** [comment]
Thanks ! 
A fully featured console could be interesting down the line: to be able to combine the same feel as the original sub-kernels' software with SoS kernel selection. But this is already fully functional with the magic words so this is probably good for now !
